### PR TITLE
composit orchestration

### PR DESCRIPTION
**Pull request by mrmasterplan**
### Principles

All ETL classes, **Orchestrator**, **Extractor**, **Transformer** and **Loader** are ETL objects.
This means that they have a method `etl(self, inputs: dataset_group) -> dataset_group`
(where `dataset_group = Dict[str, DataFrame]`) that transforms a set of import to a set of 
outputs. The special properties of each type are
 - **Extractor** always adds its result to the total set,
 - **Transformer** consumes all inputs and adds a single result dataframe,
 - **Loader** acts as a sink, while passing its input on to the next sink,

The special case of the  **Orchestrator** is that it takes all its steps and executes them
in sequence on its inputs. Running in the default `execute()` method, the inputs are empty,
but an orchestrator can also be added as part of another orchestrator with the `step` method.

For the most general case of a many to many transformation, implement your step by inheriting
from the `EtlBase` class.

### CODE REVIEW

**Review by christianhelle — COMMENTED** [review]
Isn't replacing pass-through virtual function implementations with `raise NotImplementedError()` is a breaking change? I'm not actively working with data or this library so I of course don't really mind, but others who depend on this could be affected

**Comment by JeppeBlixen** [comment]
This looks as a great solution, on the issue of having multiple orchestrator layers. I think it would be hreat to have an example og this, both for testing and documentation

**Comment by mrmasterplan** [comment]
I changed this a bit more. `execute` is now an alias for `etl`. 
To really support composit orchestration for now the extractor needs to handle the case that a key was already given. It this is not handled, the transformer may fail. That makes for a terrible error message. I therefore added  some logic to give helpful warnings if you make this exact mistake.

File: src/atc/etl/orchestrator.py [hunk 1]
```diff
@@ -24,7 +25,38 @@ def step(self, etl: EtlBase) -> "Orchestrator":
...
+        inputs = inputs or {}
+
+        # make a shallow copy of the inputs for waring after the first step
+        datasets = inputs.copy()
+        if not self.steps:
+            raise NotImplementedError("The orchestrator has no steps.")
+
+        # treat the fist step differently to warn in case the input was not handled
+        datasets = self.steps[0].etl(datasets)
+        if len(inputs) and (len(inputs) + 1 == len(datasets)):
```
Comment (JeppeBlixen) [inline src/atc/etl/orchestrator.py:38]: I do not understand this need to check if the fist step have cleaned up the inputs.
I could imagine a valid scenario where i have an input dataframe, a first step extractor reading a dataframe, and a transformer working on both of them?

File: src/atc/etl/extractor.py [hunk 2]
```diff
@@ -15,6 +15,7 @@ class Extractor(EtlBase):
     """
 
     def __init__(self, dataset_key: str = None):
+        super().__init__()
```
Comment (JeppeBlixen) [inline src/atc/etl/extractor.py:18]: This is maybe a detail, but all these __init__ updates to call the parent constructors. Should these be in its own PR? It does not feel like its related composit-orchestration.

File: src/atc/etl/orchestrator.py (hunk 1 above)
Comment (mrmasterplan) [reply src/atc/etl/orchestrator.py:38]: Absolutely, this is a special case which is why it is only a warning. Do we need a flag to disable the warning?

File: src/atc/etl/extractor.py (hunk 2 above)
Comment (mrmasterplan) [reply src/atc/etl/extractor.py:18]: they are generally good practice I believe, and do no harm.

File: src/atc/etl/orchestrator.py (hunk 1 above)
Comment (mrmasterplan) [reply src/atc/etl/orchestrator.py:38]: I used the warnings package so that it is only raised the first time. In additon, I added a suppression flag.

**Review by JeppeBlixen — APPROVED** [review]
